### DISCUSSION TITLE
file attachment

### DISCUSSION THREAD

**Discussion by JoeyZhouAus** [umputun/remark42#1253] [General]
Hi guys, do we have any support for file attachment on top of images?
Thanks!

**Comment by paskal** [comment]
We don't have support for uploading anything but images to your comments, and I think embedding a file hosting into Remark42 is not a good idea for multiple reasons.

Users can easily solve the problem of uploading the file alongside the comment using one of the numerous online services created to share files.

**Reply by JoeyZhouAus** [reply]
Hi @paskal , Thanks for getting back to me. Remark42 is really impressive and great looking product compared to ISSO and other alternatives, especially it has a built-in Image upload support.

 I am trying to experiment it. will it be difficult for me to customise it to allow pdf upload capability into it?
 or as you have mentioned, what kind of online file sharing services can be embedded alongside the comment? because as far as i can see, the remark42 is a standalone drop-in kit by itself.
 Thanks!

**Reply by paskal** [reply]
If you have a use-case where your users frequently need to upload something and share it in comments, I think the simplest solution is adding a separate instructions block on top of the comments form, something like "Use https://example.org for sharing the files in the comments".

It would be troublesome to integrate it into remark42 itself and maintain a fork. If you want to try, the simplest way to go is to hack [savePictureCtrl](https://github.com/umputun/remark42/blob/master/backend/app/rest/api/rest_private.go#L696-L718) to allow PDFs into image storage.

**Reply by JoeyZhouAus** [reply]
Cool, Thanks @paskal

**Comment by umputun** [comment]
I do agree with @paskal - storing any type of attached file in remark42 seems to be outside of the scope of the commenting system and external web storage system with links will be the better solution.

Regarding pdf files - theoretically, we can consider those as some sort of embeddable content and handle it the same way we handle pictures, however from UI point of view this going to be different. Likely it will need an iframe with embedded pdf view to show the preview version, or some sort of "attachment" view. 

Generally speaking, I have nothing against adding pdf support, probably this will be relatively easy change on the backend side but much more work on the frontend side. @JoeyZhouAus if you want to make it happen, all yours and PRs are welcome.

**Reply by JoeyZhouAus** [reply]
yeah, Thanks @umputun, i was experimenting with it, and managed to drop a pdf and retrieved it as well, now i am up to the point to make the retrieved pdf as a downloadable link, i will see how it goes, cheers

**Comment by JoeyZhouAus** [comment]
Hi @umputun  and @paskal , 
I am having some trouble of understanding how does the image get loaded in the frontend, i found the code snippet that is relevant to image loading as:
`            <div
              className={b('comment__text', { mix: b('raw-content', {}, { theme: props.theme }) })}
              // eslint-disable-next-line react/no-danger
              dangerouslySetInnerHTML={{ __html: o.text }}
            />`
in file comment.tsx, please correct me if i am wrong, but how does the above code get turned into a div that contains an image embedded into a <p> section?
![image](https://user-images.githubusercontent.com/22076698/155627370-a4e7f460-9c20-452f-bc82-1ce9adf7a4b0.png)

Thanks

**Reply by paskal** [reply]
I'm not an expert on frontend and don't know the answer to your question, unfortunately.

**Reply by JoeyZhouAus** [reply]
Hi @paskal, seems like the <img annotation of the link is prepared in the GoLang side using a markdown processor through another library:
![image](https://user-images.githubusercontent.com/22076698/156462677-3f323951-430e-455d-8fb6-9ad4f8f571bd.png)

do you know where do we prepare the node type for the attachment, so i can change the node type accordingly depends on it is a picture or pdf, thus when it comes down to the switch case again, for pdf it will be constructed with <a anchor annotation, and for image it will stay the way it is now.

Btw, this is the RenderNode function i was talking about above
![image](https://user-images.githubusercontent.com/22076698/156463096-9052e722-4353-4564-b487-b19c3b787a4f.png)


Thanks

**Reply by paskal** [reply]
Could you please write to me [in Telegram](https://t.me/paskal_07)? I think it would be easier for me to help you that way.
